lib/rex/proto/kerberos/pac/krb5_pac.rb

```diff
@@ -0,0 +1,406 @@
+    krb5_logon_information 0x00000001
+    krb5_client_info 0x0000000A
+    krb5_pac_server_checksum 0x00000006
+    krb5_pac_priv_server_checksum 0x00000007
+  end
+
+  class Krb5PacInfoBuffer < BinData::Record
+    endian :little
+
+    # @!attribute [rw] ul_type
+    #   @return [Integer] Describes the type of data present in the buffer
+    uint32 :ul_type
+
+    # @!attribute [rw] cb_buffer_size
+    #   @return [Integer]
+    uint32 :cb_buffer_size, initial_value: -> { buffer.pac_element.num_bytes }
+
+    # @!attribute [rw] offset
+    #   @return [Integer]
+    uint64 :offset
+
+    delayed_io :buffer, read_abs_offset: :offset do
+      # @!attribute [rw] pac_element
+      #   @return [Krb5PacElement]
+      krb5_pac_element :pac_element, selection: -> { ul_type }
+      string :padding, length: -> { num_bytes_to_align(pac_element.num_bytes) }
+    end
+  end
+
+  class Krb5Pac < BinData::Record
+    endian :little
+    auto_call_delayed_io
+
+    # @!attribute [rw] c_buffers
+    #   @return [Integer]
+    uint32 :c_buffers, asserted_value: -> { pac_info_buffers.length }
+
+    # @!attribute [r] version
+    #   @return [Integer]
+    uint32 :version, asserted_value: 0x00000000
+
+    # @!attribute [rw] pac_info_buffers
+    #   @return [Array<Krb5PacInfoBuffer>]
+    array :pac_info_buffers, type: :krb5_pac_info_buffer, initial_length: :c_buffers
+
+    def assign(val)
+      case val
+      when Hash
+        pac_infos = val[:pac_elements].map do |pac_element|
+          { ul_type: pac_element.ul_type, buffer: { pac_element: pac_element } }
+        end
+        new_val = val.merge(pac_info_buffers: pac_infos)
+        super(new_val)
+      else
+        super
+      end
+    end
+
+    # Calculates the checksums, can only be done after all other fields are set
+    def calculate_checksums(key: nil)
+      server_checksum = nil
+      priv_server_checksum = nil
+      pac_info_buffers.each do |info_buffer|
+        pac_element = info_buffer.buffer.pac_element
+        if pac_element.ul_type == 6
+          server_checksum = pac_element
+        elsif pac_element.ul_type == 7
+          priv_server_checksum = pac_element
+        end
+      end
+      server_checksum.signature = calculate_checksum(server_checksum.signature_type, key, to_binary_s)
+
+      priv_server_checksum.signature = calculate_checksum(priv_server_checksum.signature_type, key, server_checksum.signature)
+    end
+
+    # Calculates the offsets for pac_elements if they haven't yet been set
+    def calculate_offsets
+      offset = pac_info_buffers.abs_offset + pac_info_buffers.num_bytes
+      pac_info_buffers.each do |pac_info|
+        next unless pac_info.offset == 0
+
+        pac_info.offset = offset
+        offset += pac_info.cb_buffer_size
+        offset += num_bytes_to_align(offset)
+      end
+    end
+
+    # Call this when you are done setting fields in the object
+    # in order to finalise the data
+    def finish(key: nil)
```

Comment:
Maybe one for group review; I'm not a super big fan of this API

If we made the mutation more obvious - it might be slightly better? 🤔 

```ruby
def sign!(key: nil)
  calculate_offsets!
  calculate_checksums!(key: key)
end
```

Rationale: Normally a method name such as `calculate_` would return a value - and not cause a side effect

Reply:
yea agree, I don't love it, ideally I'd like for bindata to handle this for us, it seems like it should be possible but I just couldn't figure out how to get it to fill in these values